_posts/2025-02-20-agentic-ai-with-quarkus-p2.adoc

```diff
@@ -0,0 +1,312 @@
+	String extractCity(String question);
+}
+----
+
+2 .  a https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/aiastool/geo/GeoCodingService.java[web service] returning the geographic coordinates of a given city.
+
+[source, java]
+----
+@RegisterRestClient(configKey = "geocoding")
+@Path("/v1")
+public interface GeoCodingService {
+
+	@GET
+	@Path("/search")
+	@ClientQueryParam(name = "count", value = "1")
+	@Tool("Finds the latitude and longitude of a given city")
+	GeoResults findCity(@RestQuery String name);
+}
+----
+
+
+3 . a https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/aiastool/weather/WeatherForecastService.java[second web service] providing the weather forecast for the given latitude and longitude.
+
+[source, java]
+----
+@RegisterRestClient(configKey = "openmeteo")
+@Path("/v1")
+public interface WeatherForecastService {
+
+	@GET
+	@Path("/forecast")
+	@Tool("Forecasts the weather for the given latitude and longitude")
+	@ClientQueryParam(name = "forecast_days", value = "7")
+	@ClientQueryParam(name = "daily", value = {
+        	"temperature_2m_max",
+        	"temperature_2m_min",
+        	"precipitation_sum",
+        	"wind_speed_10m_max",
+        	"weather_code"
+	})
+	WeatherForecast forecast(@RestQuery double latitude, @RestQuery double longitude);
+}
+----
+
+.Weather agent architecture
+image::weather-agent.png[]
+
+It's possible to invoke the https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/aiastool/WeatherResource.java[HTTP endpoint] exposing this agent-based weather service, for example asking it what's the weather in Rome:
+
+[source, shell]
+----
+curl http://localhost:8080/weather/city/Rome
+----
+
+The response will be something like:
+
+----
+The weather in Rome today will have a maximum temperature of 14.3°C, minimum temperature of 2.0°C.
+No precipitation expected, and the wind speed will be up to 5.6 km/h.
+The overall weather condition is expected to be cloudy.
+----
+
+In essence, this control flow is quite similar to what is described in the prompt chaining workflow, where the user input is sequentially transformed in steps (in this case, going from the prompt to the name of the city contained in that prompt to the geographical coordinates of that city, to the weather forecasts at those coordinates). The significant difference is that the LLM directly orchestrates the sequence of steps instead of being externally orchestrated programmatically.
+
+=== A more general purpose AI agent
+
+An AI agent can have even more freedom to operate, thus being more flexible and covering a wider range of tasks. Typically, these scenarios requires a bigger model, so https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/searchastool[this second example] has been configured with `qwen2.5-14b` and a longer timeout to give it a chance to complete its task.
```

Comment:
would be good to introduce this section with a sentence describing the scenario that you're about to provide the technical details for.